aiida/cmdline/utils/migration/v03_to_v04.py

```diff
@@ -0,0 +1,452 @@
+        if node['type'].startswith('calculation.job.'):
+            node['type'] = 'node.process.calculation.calcjob.CalcJobNode.'
+
+        # update type for InlineCalculation nodes
+        if node['type'] == 'calculation.inline.InlineCalculation.':
+            node['type'] = 'node.process.calculation.calcfunction.CalcFunctionNode.'
+
+        # update type for FunctionCalculation nodes
+        if node['type'] == 'calculation.function.FunctionCalculation.':
+            node['type'] = 'node.process.workflow.workfunction.WorkFunctionNode.'
+
+    uuid_node_type_mapping = {node['uuid']: node['type'] for node in data['export_data']['Node'].values()}
+    for link in data['links_uuid']:
+        inp_uuid = link['output']
+        # rename `createlink` to `create`
+        if link['type'] == 'createlink':
+            link['type'] = 'create'
+        # rename `returnlink` to `return`
+        elif link['type'] == 'returnlink':
+            link['type'] = 'return'
+
+        elif link['type'] == 'inputlink':
+            # rename `inputlink` to `input_calc` if the target node is a calculation type node
+            if uuid_node_type_mapping[inp_uuid].startswith('node.process.calculation'):
+                link['type'] = 'input_calc'
+            # rename `inputlink` to `input_work` if the target node is a workflow type node
+            elif uuid_node_type_mapping[inp_uuid].startswith('node.process.workflow'):
+                link['type'] = 'input_work'
+
+        elif link['type'] == 'calllink':
+            # rename `calllink` to `call_calc` if the target node is a calculation type node
+            if uuid_node_type_mapping[inp_uuid].startswith('node.process.calculation'):
+                link['type'] = 'call_calc'
+            # rename `calllink` to `call_work` if the target node is a workflow type node
+            elif uuid_node_type_mapping[inp_uuid].startswith('node.process.workflow'):
+                link['type'] = 'call_work'
+
+
+def migrate_v3_to_v4(metadata, data, folder, *args):
+    """
+    Migration of export files from v0.3 to v0.4
+    """
+    old_version = '0.3'
+    new_version = '0.4'
+
+    verify_metadata_version(metadata, old_version)
+    update_metadata(metadata, new_version)
+
+    # Apply migration: 0009 - REV. 1.0.9
+    # `DbNode.type` content changes:
+    # 'data.base.Bool.'  -> 'data.bool.Bool.'
+    # 'data.base.Float.' -> 'data.float.Float.'
+    # 'data.base.Int.'   -> 'data.int.Int.'
+    # 'data.base.Str.'   -> 'data.str.Str.'
+    # 'data.base.List.'  -> 'data.list.List.'
+    if 'Node' in data['export_data']:
+        for content in data['export_data']['Node'].values():
+            if 'type' in content and content['type'].startswith('data.base.'):
+                type_str = content['type'].replace('data.base.', '')
+                type_str = 'data.' + type_str.lower() + type_str
+                content['type'] = type_str
+
+    # Apply migrations: 0010 - REV. 1.0.10
```

Comment:
unless they're super short, most of these migrations should probably be separate functions, would you agree?

Reply:
It would be the most ideal and "pythonic" way, I guess. I will gladly do it, if you think it will result in a more transparent code in the end?

Reply:
> if you think it will result in a more transparent code in the end

Yes. Linters complain about too long/complex functions not just to make programmers miserable but because there are good reasons to keep them short.

One example is just variable scope - in a function that's hundreds of lines long it can be very difficult to figure out which parts touch which variables.
Isolating individual steps into functions (you've already done most of the job in the comments) makes this much clearer.

Reply:
Thank you. Will make this a priority.

Reply:
Done